docs/userGuide/syntax/lists.md

```diff
@@ -109,3 +109,126 @@ first number
 - [x] Item 4
 - ( ) Item 5
 </div>
+
+****Customizing the list appearance:****
+<box type="tip" seamless>
+The list icons can be customized by adding an {icon="..."} at the end of a list item or a list itself.
+</box>
+
+
+<div id="main-example-markbind">
+
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">markdown</variable>
+<variable name="code">
+* item 1 {icon="x"}
+* item 2 {icon="heavy_check_mark"}
+* item 3
+* item 4 {icon="/images/deer.jpg"}
+  
+{icon="fas fa-check"}
+</variable>
+</include>
+
+</div>
+
+<box type="warning">
+
+1. Item-level specifications should only be attached to the list heading or list item, not the list content.
+2. A blank line preceding the content is crucial.
+3. It's important to insert a blank line before list-level specifications.
+4. It is possible to any of the icons supported by MarkBind (e.g., Font Awesome, Octicons, Glyphicons, Images and etc.) as the icon of a list item.
+5. Images will maintain a 1 rem bottom margin.
+</box>
+
+<box type="tip" seamless>
+It is possible to specify icons for nested levels using the icons attribute.
+</box>
+<div id="main-example-markbind">
+
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">markdown</variable>
+<variable name="code">
+* item 1
+* item 2 
+  * item 2.1
+  * item 2.2
+  {icon="x"}
+* item 3
+  
+{icon="fas fa-check"}
+</variable>
+</include>
+
+</div>
+
+<box type="tip" seamless>
+The icon appearance can be customized further a class attribute.
+</box>
+<div id="main-example-markbind">
+
+<include src="codeAndOutput.md" boilerplate >
+<variable name="highlightStyle">markdown</variable>
+<variable name="code">
+* item 1 {icon="fas fa-check" class="text-warning"}
+* item 2 {icon="fas fa-check" class="text-danger"}
+* item 3 
+  
+{icon="fas fa-check" class="text-primary"}
+</variable>
+</include>
+
+</div>
```

Comment:
Note that icons (just like characters) have different widths, even if the size is the same. However, we want all items/icons in the list to be aligned nicely, if the icons are of same size. Otherwise the list will look very untidy (like the example below -- I used letters instead of icons).

**X**  item 1
**I**  item 2
**W**  item 3 

When aligned properly:
`X`  item 1
`I`  item 2
`W`  item 3

Reply:
The following has been taken into account:
 
<img width="437" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/cb72639e-c067-4ed2-88be-5000822ee713">

Reply:
I see. Should be fine if you also considered that the widths of characters differ even at the same font size (same for font-based icons such as fontawesome).
Anyway, you can test with something like this, to see if the outcome is aligned nicely:
```
* item 1 { icon="fas home" } 
* item 1 { icon="fas trophy" } 
* item 1 { icon="glyphicon education" } 
 ```

Reply:
Here is the sample output:
<img width="503" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/1f3197b3-0b68-48f0-8e63-9467f71ed785">

Reply:
Looks good. Seems that problem has been taken care of already 👍

Reply:
Imo, fixed sizes are user friendly but I often find myself wanting to do more fine grained adjustments and it ends up being less so.

How about:
- Icon used: `height=width` by default (as icons tend to be square), `width=1em` (responsive to text size)
- Image used: `height=50px` (or anything else that makes sense), let `width` adjust according to the image.
- Support specifiying any units in `icon-width + icon-height` attributes 

In consideration with the below comment on aspect ratios too*

Reply:
https://github.com/MarkBind/markbind/pull/2316#discussion_r1252088583

Reply:
Another thing if you have the time, you can consider is providing an alignment option for the case that icon widths are different (e.g. images of different sizes); But I think this is rarer and less important for now, we can pick a good default and add it later if needed.

Reply:
Imo excessive freedom doesn't always yield positive results. Although an image can have a 16:9 ratio or non-square dimensions, it's easy to envision that such an image wouldn't work well as an icon and may not be aesthetically pleasing. Nonetheless, it's still worth considering.

Reply:
if you mean aesthetics, it can be fairly pleasing:

![image](https://github.com/MarkBind/markbind/assets/3306138/4b92a131-2706-4a47-9aa1-2bfecf9c8e21)

it'd be up to the author to use something appropriate, we merely provide that flexibility of non-square images.

Reply:
> Image used: width=50px (or anything else that makes sense), let height adjust according to the image.

If the author's original image were already square, it should also work out nicely without extra configuration.

Reply:
> We can indeed let users define aspect ratio

Just so we're on the same page I'm not suggesting an `icon-aspect-ratio` of sorts, instead authors should be able to let the aspect ratio be defined by their media size (while being able to specify height, or width).

Similar to how you would configure a regular `<img>`

> Restrict setting specific width and height aims to prevent unpredictable displays, such as a 500px vs 30px scenario.

Not sure I understood this, could you clarify?

> Also, we need to consider mobile view display.

I don't think this is something we can do anything much about in either case. Unless we do something drastic like placing the icon above the content which seems like a bad idea.

Again, similar to how you would think of putting a regular `<img>` on mobile, it is up to the author to size it appropriately.

Some minor "solutions" I can think of currently otherwise:
- drop a well placed documentation note/tip on this.
- Maybe also tighten our existing margins/paddings.
- something like `icon-mobile-size/icon-mobile-width` but I'm not sure if its necessary for now, we can address it if the need pops up. (considering you already added a `icon-class` feature too)

Reply:
> Just so we're on the same page I'm not suggesting an `icon-aspect-ratio` of sorts, instead authors should be able to let the aspect ratio be defined by their media size (while being able to specify height, or width).

This is just my design consideration.

> Not sure I understood this, could you clarify?

For instance, consider iPhone widgets. Developers can only create widgets of specific sizes. This restriction enables users to have an optimal display, irrespective of the size and device they use.

> * something like `icon-mobile-size/icon-mobile-width` but I'm not sure if its necessary for now, we can address it if the need pops up. (considering you already added a `icon-class` feature too)

I don't think that 'icon-mobile-size' or 'icon-mobile-width' are good suggestions. Imo, all these tedious specifications should be handled by the framework to optimize UX.
The width and height should be the most lenient we can go. They should be sufficient for our use case.

Reply:
Sorry, I'm still not quite sure what you are suggesting (if any). To clarify, are you good with this approach as well? https://github.com/MarkBind/markbind/pull/2316#discussion_r1252094010

Or are you suggesting we keep the existing square approach, or something else entirely.

> The width and height should be the most lenient we can go. They should be sufficient for our use case.

Yes, it should be.

> Imo, all these tedious specifications should be handled by the framework to optimize UX.

Do you mean we should go even further with something like this?

1. Allow the author to only control `icon-width=Xpx`

1. Don't trust the author to do the right thing:

```css
@media only screen and (max-width: 768px) {
  .icon {
    /* For mobile, it should not be too big nor small */
    min-width: 20px;
    max-width: 100px;
  }
}
```

Reply:
> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

I agree this should take priority over transparency.

> Do you mean we should go even further with something like this?

What do you think of this? (as a whole) 

- Configurable `icon-width` in all cases.
- Icon used: `height=width` mandatorily, `width=some default`
- Images/svg used: aspect ratio will be mandatorily preserved.
- `max-width: 100px` (feel free to adjust the number) on some mobile breakpoint
   - `min-width` may not be necessary as the author will notice the problem even on non-mobile devices

If we have a need for `icon-height` down the line we can add it

Reply:
Here is my suggestions:
Size: square (default to minimal if no size specified),
Optional: width (overrides default),
Optional: height (overrides default).

"Max-width/height: 100px" is a redundant attribute. If users are mindful about this, display issues won't be a concern. However, we can still include it as an optional attribute.

Reply:
> Size: square

I agree on icons but strongly think images/svg/media should use their native aspect ratio by default. If the image was already meant to be square - it'll stay that way. Rectangular/portrait images should stay that way as well.


> "Max-width/height: 100px" is a redundant attribute. If users are mindful about this, display issues won't be a concern. However, we can still include it as an optional attribute.

Dosen't this go against ⬇️?

> > Don't trust the author to do the right thing:
>
> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

Reply:
> I agree on icons but strongly think images/svg/media should use their native aspect ratio by default. If the image was already meant to be square - it'll stay that way. Rectangular/portrait images should stay that way as well.

If square media icons are preferred - the author should crop it appropriately using some external software to preserve an unskewed image. Think the circular crop UI you see around profile pages - though we don't have that luxury in markdown (imo it'd be too difficult to use/implement in markdown)

Reply:
> "Max-width/height: 100px" is a redundant attribute

Not as an attribute to be clear, but the CSS we talked about earlier:

> > Don't trust the author to do the right thing:
> > ```css
> > @media only screen and (max-width: 768px) {
> >   .icon {
> >     /* For mobile, it should not be too big nor small */
> >     min-width: 20px;
> >     max-width: 100px;
> >   }
> > }
> > ```
>
> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

Reply:
> > Imo this is what we should aim for, as md should be easy to use with minimal consideration.

I see your point, but it might confuse users when their height/width settings don't work, because they didn't specify the min or max attributes.

Reply:
Not as a HTML attribute at all; But CSS.

> Imo this is what we should aim for, as md should be easy to use with minimal consideration.

Isn't this your point? *quite confused* 😂

Yes, I've mentioned transparency earlier. On second thought (with your point on easy to use), not trusting the author should take priority.

The breakpoint dosen't have to be `768px` as well, something even smaller like `576px` could work. Similarly, the `max-width` dosen't  have to be `100px`, maybe something like `200px`. At the point it would fill nearly half the screen or more (depending on the preexisting indentation of the list) and we can quite confidently question the author's intention).

Reply:
> Isn't this your point? _quite confused_ 😂

In my opinion, we shouldn't make users feel restricted or surprised by hidden boundaries like a maximum width. Setting up sizes such as XS, S, M, L, XL, instead of width and height with an invisible maximum, could meet user requirements more intuitively. This approach would avoid the frustration of configuring width and height, hitting unexpected limits, and then having to troubleshoot by referring to notes and FAQs in the docs.

Nonetheless, the provided width and height should adequately support the basic usage of all image types within the list.

> The breakpoint dosen't have to be `768px` as well, something even smaller like `576px` could work. Similarly, the `max-width` dosen't have to be `100px`, maybe something like `200px`. At the point it would fill nearly half the screen or more (depending on the preexisting indentation of the list) and we can quite confidently question the author's intention).

However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

Reply:
> I am good width width and height suggestions.

I see, so by this you meant we should have `width` + `height` but still stick with fixed sizes to eliminate the possibility for users to construct an inappropriate size on mobile. (correct me if I'm wrong)

Here's my original issue with this approach's UX:

> Imo, fixed sizes are user friendly but I often find myself wanting to do more fine grained adjustments and it ends up being less so.

- `l=70px`
- `xl=96px`

When I'm adjusting media/icons, I often find myself, especially in the case of media, wanting to fine tune things (e.g. to `80px`) to look nice. I end up having to use an unwieldy `style=...` (not even possible here) or define a whole other css class to stick to `icon-class` which is rather inconvenient and counter to having a `size/width/height` attribute.

To fix this I can think of the following approaches:

1. Vuetify for example, https://vuetifyjs.com/en/api/v-icon/#props-size, supports predefined sizes, but also finer grained adjustments with px. I'm fine with following this too if you strongly believe having `xs,s,...` alongside would be worth it.
2. Support only the form that can adequately handle all use cases (specific widths and heights in `px`) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a `px` is, or be able to intuitively bump up or down a number controlling the icon size.

As we discussed there's also an issue here of authors specifying too large a width on mobile and things going haywire. If we would like to "fix" this, we could:

0. Don't fix it at all, trust the author

1. Specifiy `max-width` on mobile devices

   > In my opinion, we shouldn't make users feel restricted or surprised by hidden boundaries like a maximum width.

   Transparency is indeed an issue, but I'd ask a few more questions as this is a tradeoff: Would having a list item icon fill half the screen on mobile device be a reasonable use case? And how much should we value supporting such use cases over authors not controlling widths properly. (imo, something generous like `max-width: 250px` on `576px` breakpoint could work to break this tie)

2. > However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

   What do you think of https://github.com/MarkBind/markbind/pull/2316#discussion_r1254701627? Is this something like what you have in mind?

Reply:
> I see, so by this you meant we should have `width` + `height` but still stick with fixed sizes to eliminate the possibility for users to construct an inappropriate size on mobile. (correct me if I'm wrong)

I'm not sure if you prefer having sizes or not - I'm fine with either. All of the above are merely suggestions. The width and height options seem to provide the best flexibility for fine-tuning.

> 1. Vuetify for example, https://vuetifyjs.com/en/api/v-icon/#props-size, supports predefined sizes, but also finer grained adjustments with px. I'm fine with following this too if you strongly believe having `xs,s,...` alongside would be worth it.

I don't particularly advocate for having sizes like 'xs', 's', etc., as shown in your example. It's indeed an interesting approach that other frameworks are implementing. Naturally, we always have the option to define the behavior of the component ourselves, which is one of the wonderful aspects of programming.

> Support only the form that can adequately handle all use cases (specific widths and heights in px) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a px is, or be able to intuitively bump up or down a number controlling the icon size.

That should be quite true.

> As we discussed there's also an issue here of authors specifying too large a width on mobile and things going haywire. If we would like to "fix" this, we could:
> 
> 0. Don't fix it at all, trust the author
> 1. Specifiy `max-width` on mobile devices
>    > In my opinion, we shouldn't make users feel restricted or surprised by hidden boundaries like a maximum width.
>    
>    
>    Transparency is indeed an issue, but I'd ask a few more questions as this is a tradeoff: Would having a list item icon fill half the screen on mobile device be a reasonable use case? And how much should we value supporting such use cases over authors not controlling widths properly. (imo, something generous like `max-width: 250px` on `576px` breakpoint could work to break this tie)
> 2. > However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.
>    
>    
>    What do you think of [Add Icon and Media Support to List  - Resolves #906 and #2272 #2316 (comment)](https://github.com/MarkBind/markbind/pull/2316#discussion_r1254701627)? Is this something like what you have in mind?

I strongly oppose converting the list behavior to a flex view. In my opinion, a list should function as a list unless specified otherwise.

I have done some experiance with different size unit:
```
    width: 50vw;
    height: 50vw;
```
Output:
<img width="658" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/6eee3804-ec60-4daf-bbfb-768d37e61d63">
<img width="319" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/f50e2532-7ab7-451f-b46c-7a1b950f4638">

```
width: 400px;
height: 400px;
```
<img width="673" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/ea61cf83-927b-4705-9591-5b0a7bc548ee">
<img width="456" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/1403bd85-dc77-47ce-9b10-94faa71a91e4">


As demonstrated by the output, using 'vw' (viewport width) is highly effective in managing the size across different screens.

Perhaps, rather than restricting size to px, we could allow users to utilize any size unit. Whether they use it correctly or not, we can leave that responsibility to them.

Reply:
> I'm not sure if you prefer having sizes or not - I'm fine with either. All of the above are merely suggestions. The width and height options seem to provide the best flexibility for fine-tuning.

Do you mean having a set fixed sizes? No, I don't prefer having _only_ it as I often find myself wanting to do fine grained adjustments. (see above [example](https://github.com/MarkBind/markbind/pull/2316#discussion_r1255844469)) But having it _alongside_ other css units (`px/vw/etc.`) is fine.

If you mean a single square `size` vs 2 `width/height` attributes, I strongly prefer `width/height` as media should be scaled according to its native aspect ratio to avoid distortion.

---

> I strongly oppose converting the list behavior to a flex view. In my opinion, a list should function as a list unless specified otherwise.

> It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

Did you have something different in mind?

---

Which of all the above approaches do you prefer currently?

https://github.com/MarkBind/markbind/pull/2316#discussion_r1255844469

Personally I'm leaning toward `2` (just css units) and then either `0` (don't enforce max-width at all) or `1` (enforce a _generous_ max-width on mobile).

Reply:
>Personally I'm leaning toward 2 (just css units) and then either 0 (don't enforce max-width at all) or 1 (enforce a generous max-width on mobile).

Given that there are no strong opinions on this, I suggest we initially go with 0. As demonstrated above, users can choose different size units to accommodate various scenarios and screen sizes.

Reply:
Sounds good, we can add on `1` later as well if we change our minds.

Reply:
> > Personally I'm leaning toward 2 (just css units) and then either 0 (don't enforce max-width at all) or 1 (enforce a generous max-width on mobile).
> 
> Given that there are no strong opinions on this, I suggest we initially go with 0. As demonstrated above, users can choose different size units to accommodate various scenarios and screen sizes.

so: `2` + `0` (cmiiw @SPWwj)

Reply:
Only '0', not '2'. What you suggested is something like a card view rather than a list.

Reply:
not the later `2` but the earlier one.

> Support only the form that can adequately handle all use cases (specific widths and heights in px) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a px is, or be able to intuitively bump up or down a number controlling the icon size.

Reply:
>Support only the form that can adequately handle all use cases (specific widths and heights in px) for feature simplicity. I'm assuming that someone using a static site generator would have at least a brief idea of what a px is, or be able to intuitively bump up or down a number controlling the icon size.

Yes, I will go forward with above.

Just to share a trick you can use to achieve what you described - arranging images and text in separate rows on a mobile screen.

 ```
* :glyphicon-education: item 1x {icon=" "}
* ![Cardview](https://github.com/SPWwj/Spirify-Ui/blob/master/src/assets/images/sharing.png?raw=true){width="350px" height="100px"} Hello world
  
  Hello world
```

<img width="444" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/7693a24a-a495-4614-a0f5-976e60c39c50">


<img width="243" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/2d83bacf-06b2-4e57-bea5-fa872a47e58b">

Reply:
Makes sense, wont need it to be in the icon t all.

> I strongly oppose converting the list behavior to a flex view. In my opinion, a list should function as a list unless specified otherwise.

> It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

> Did you have something different in mind?

Was this layout (top and bottom) what you meant by `flex` as well? I cannot imagine how else you would position the icon and image (top and bottom vs left and right).

Reply:
> Was this layout (top and bottom) what you meant by flex as well? 

Refer to this: https://www.w3schools.com/css/css3_flexbox_responsive.asp

> I cannot imagine how else you would position the icon and image (top and bottom vs left and right).

The behavior of the list is oriented towards the left (icon) and right. So, simply use the icon attribute. If you prefer a top and bottom arrangement, remove the icon and focus on the content. This will nicely arrange items top and bottom on smaller screen sizes.

Reply:
> Refer to this: https://www.w3schools.com/css/css3_flexbox_responsive.asp

Yes, but how would you use this to achieve something different from "top and bottom vs left and right"?

Reply:
Not sure what you mean?

Reply:
> However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

What's the UI layout for the icon and list item content you were thinking of here?

Reply:
This represents a list:
Icon | Content
Icon | Content
Icon | Content

However, the following represents a card view layout similar to [Ant Design Cards](https://ant.design/components/card):
Icon
Content
Icon
Content
Icon
Content

Switching from a list to a card view layout may result in unexpected alignment issues. Imagine a scenario with a small icon and extensive text:

On a large screen, it might look like this:
Small icon: very long text repeated numerous times

But on a smaller screen, the alignment might change dramatically:
Small icon:
Very long text repeated numerous times

Clearly, the change in layout can cause significant alignment issues.

In addition, by maintaining the layout of a standard list, we can eventually achieve a UI layout like this:
<img width="408" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/abb71d6e-0807-441b-b664-a6789ad9b914">

Moreover, currently, we are crafting an icon list. However, in the future, you might need a navigation list, a grid list, and so forth. Hence, it's advisable to follow the standard list format unless specified otherwise by an attribute.

Reply:
Sorry, I've only just gotten around to reading the `flex` part of your code.

There's likely some miscommunication here.

> However, such an approach could disrupt the list presentation, which may not be ideal. It would be more effective to specify an attribute like {flex} to achieve the desired presentation.

I understood from the conversation flow that you are suggesting `flex` as a solution to the user restrictiveness imposed by `max-width`. I don't see how this can be done apart from adopting a completely different UI layout. (with `flex` as the implementation)

I think we are talking about the same thing - just whether having `max-width` or not will be ideal in terms of just sticking to `flex` in your current implementation, inside a conventional "icon on left content on right" layout.

I'll resolve this but let me know if I got this wrong.